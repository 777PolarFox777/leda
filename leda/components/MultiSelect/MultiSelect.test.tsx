import React from 'react';
import userEvent from '@testing-library/user-event';
import {
  render, screen,
} from '@testing-library/react';
import { MultiSelect } from './index';

describe('MultiSelect SNAPSHOTS', () => {
  it('should render basic usage', () => {
    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value1']} />
    ));

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render value in controllable mode', () => {
    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} />
    ));

    expect(document.querySelector('span.tags-item')).toHaveTextContent('value0');

    expect(wrapper.container).toMatchSnapshot();

    wrapper.rerender((
      <MultiSelect data={['value0', 'value1']} value={['value1']} />
    ));

    expect(document.querySelector('span.tags-item')).toHaveTextContent('1');

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render data in controllable mode', () => {
    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} />
    ));

    expect(wrapper.container).toMatchSnapshot();

    wrapper.rerender((
      <MultiSelect
        data={[
          { id: 0, value: 'value0' },
          { id: 1, value: 'value1' },
        ]}
        value={['value1']}
      />
    ));

    expect(wrapper.container).toMatchSnapshot();
  });
});

describe('MultiSelect ATTRIBUTES', () => {
  it('should list isOpen', () => {
    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen />
    ));

    expect(document.querySelectorAll('li.suggestion-item')).toHaveLength(1);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should list loading', () => {
    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen isLoading />
    ));

    expect(document.querySelectorAll('span.loader-element')).toHaveLength(1);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should render noData template', () => {
    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen />
    ));

    expect(wrapper.container).toMatchSnapshot();

    userEvent.type(screen.getByRole('textbox'), 'clear');

    expect(document.querySelector('div.nodata')).toHaveTextContent('Ничего не найдено');

    expect(wrapper.container).toMatchSnapshot();
  });
});

describe('MultiSelect HANDLERS', () => {
  it('should test onFocus', () => {
    const handleFocus = jest.fn();

    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen name="name" onFocus={handleFocus} />
    ));

    screen.getByRole('textbox').focus();

    expect(handleFocus).toHaveBeenCalledTimes(1);

    const [[event]] = handleFocus.mock.calls;

    expect(event.component.name).toEqual('name');

    expect(event.component.value).toEqual(['value0']);

    expect(document.querySelectorAll('span.tags-item')).toHaveLength(1);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should test onBlur', () => {
    const handleBlur = jest.fn();

    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} name="name" onBlur={handleBlur} />
    ));

    expect(wrapper.container).toMatchSnapshot();

    screen.getByRole('textbox').focus();
    screen.getByRole('textbox').blur();

    expect(handleBlur).toHaveBeenCalledTimes(1);

    const [[event]] = handleBlur.mock.calls;

    expect(event.component.value).toEqual(['value0']);

    expect(event.component.name).toEqual('name');

    expect(event.component.isValid).toBeTruthy();

    expect(document.querySelectorAll('span.tags-item')).toHaveLength(1);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should test onChange', () => {
    const handleChange = jest.fn();

    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen name="name" onChange={handleChange} />
    ));

    userEvent.click(document.querySelectorAll('li.suggestion-item')[0]);

    expect(handleChange).toHaveBeenCalledTimes(1);

    const [[event]] = handleChange.mock.calls;

    expect(event.component.value).toEqual(['value0', 'value1']);

    expect(event.component.name).toEqual('name');

    expect(event.component.selectedValue).toEqual('value1');

    expect(event.component.deselectedValues).toBeUndefined();

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should test onDeselect', () => {
    const handleChange = jest.fn();

    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen name="name" onChange={handleChange} />
    ));

    userEvent.click(document.querySelectorAll('span.tags-icon')[0]);

    expect(handleChange).toHaveBeenCalledTimes(1);

    const [[event]] = handleChange.mock.calls;

    expect(event.component.value).toEqual([]);

    expect(event.component.name).toEqual('name');

    expect(event.component.selectedValue).toBeUndefined();

    expect(event.component.deselectedValues).toEqual(['value0']);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('should test onDeselect by clear button', () => {
    const handleChange = jest.fn();

    const wrapper = render((
      <MultiSelect data={['value0', 'value1']} value={['value0']} isOpen name="name" onChange={handleChange} hasClearButton />
    ));

    userEvent.click(document.querySelectorAll('span.multiselect-clear-icon')[0]);

    expect(handleChange).toHaveBeenCalledTimes(1);

    const [[event]] = handleChange.mock.calls;

    expect(event.component.value).toEqual([]);

    expect(event.component.name).toEqual('name');

    expect(event.component.deselectedValues).toEqual(['value0']);

    expect(wrapper.container).toMatchSnapshot();
  });
});
