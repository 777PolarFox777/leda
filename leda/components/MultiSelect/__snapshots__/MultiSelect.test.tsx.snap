// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect ATTRIBUTES should list isOpen 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <ul
        class="suggestion-list"
      >
        <li
          class="suggestion-item"
        >
          value1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MultiSelect ATTRIBUTES should list loading 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <div
        class="loader-wrapper"
      >
        <div
          class="loader-container"
        >
          <span
            class="loader-element"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiSelect ATTRIBUTES should render noData template 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <ul
        class="suggestion-list"
      >
        <li
          class="suggestion-item"
        >
          value1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MultiSelect ATTRIBUTES should render noData template 2`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value="clear"
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <div
        class="nodata"
      >
        Ничего не найдено
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiSelect HANDLERS should test onBlur 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        name="name"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelect HANDLERS should test onBlur 2`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        name="name"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelect HANDLERS should test onChange 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        name="name"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <ul
        class="suggestion-list"
      >
        <li
          class="suggestion-item"
        >
          value1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MultiSelect HANDLERS should test onDeselect 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper focused"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        name="name"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <ul
        class="suggestion-list"
      >
        <li
          class="suggestion-item"
        >
          value1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MultiSelect HANDLERS should test onDeselect by clear button 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper focused"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
        <span
          class="multiselect-clear-icon"
        />
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        name="name"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <ul
        class="suggestion-list"
      >
        <li
          class="suggestion-item"
        >
          value1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MultiSelect HANDLERS should test onFocus 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper focused"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        name="name"
        value=""
      />
    </div>
    <div
      class="suggestion-wrapper visible"
    >
      <ul
        class="suggestion-list"
      >
        <li
          class="suggestion-item"
        >
          value1
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MultiSelect SNAPSHOTS should render basic usage 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value1
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelect SNAPSHOTS should render data in controllable mode 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelect SNAPSHOTS should render data in controllable mode 2`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value1
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelect SNAPSHOTS should render value in controllable mode 1`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value0
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelect SNAPSHOTS should render value in controllable mode 2`] = `
<div>
  <div
    class="multiselect-wrapper"
  >
    <div
      class="multiselect-input-wrapper"
    >
      <div
        class="multiselect-tags-container"
      >
        <span
          class="tags-item"
        >
          value1
          <span
            class="tags-icon icon-default"
          />
        </span>
      </div>
      <input
        aria-invalid="false"
        class="multiselect-input"
        value=""
      />
    </div>
  </div>
</div>
`;
